src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,363 @@
+    ///
+    /// @see https://docs.mongodb.org/manual/tutorial/build-indexes-in-the-background/
+    ///
+    void background(bool background);
+
+    ///
+    /// The current background setting.
+    ///
+    /// @return The current background.
+    ///
+    const stdx::optional<bool>& background() const;
+
+    ///
+    /// Whether or not to create a unique index so that the collection will not accept insertion of
+    /// documents where the index key or keys match an existing value in the index.
+    ///
+    /// @param unique
+    ///   Whether or not to create a unique index.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-unique/
+    ///
+    void unique(bool unique);
+
+    ///
+    /// The current unique setting.
+    ///
+    /// @return The current unique.
+    ///
+    const stdx::optional<bool>& unique() const;
+
+    ///
+    /// The name of the index.
+    ///
+    /// @param name
+    ///   The name of the index.
+    ///
+    void name(std::string name);
+
+    ///
+    /// The current name setting.
+    ///
+    /// @return The current name.
+    ///
+    const stdx::optional<std::string>& name() const;
+
+    ///
+    /// Whether or not to create a sparse index. Sparse indexes only reference documents with the
+    /// indexed fields.
+    ///
+    /// @param sparse
+    ///   Whether or not to create a sparse index.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-sparse/
+    ///
+    void sparse(bool sparse);
+
+    ///
+    /// The current sparse setting.
+    ///
+    /// @return The current sparse setting.
+    ///
+    const stdx::optional<bool>& sparse() const;
+
+    ///
+    /// Optionally used only in MongoDB 3.0.0 and higher. Specifies the storage engine options for
+    /// the index.
+    ///
+    /// @param storage_options
+    ///   The storage engine options for the index.
+    ///
+    void storage_options(std::unique_ptr<base_storage_options>& storage_options);
+    void storage_options(std::unique_ptr<wiredtiger_storage_options>& storage_options);
+
+    ///
+    /// The current storage_options setting.
+    ///
+    /// @return The current storage_options.
+    ///
+    const base_storage_options* storage_options() const;
+
+    ///
+    /// Set a value, in seconds, as a TTL to control how long MongoDB retains documents in this
+    /// collection.
+    ///
+    /// @param expire_after_seconds
+    ///   The amount of time, in seconds, to retain documents.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-ttl/
+    ///
+    void expire_after_seconds(std::int32_t expire_after_seconds);
+
+    ///
+    /// The current expire_after_seconds setting.
+    ///
+    /// @return The current expire_after_seconds.
+    ///
+    const stdx::optional<std::int32_t>& expire_after_seconds() const;
+
+    ///
+    /// Sets the index version.
+    ///
+    /// @param v
+    ///   The index version.
+    ///
+    void v(std::int32_t v);
```

Comment:
I know this is idiomatically called 'v', but can we do better? This is a pretty terrible method name! Also, signed/unsigned for the parameter?